lib/ts/recipe/passwordless/index.ts

```diff
@@ -50,6 +52,167 @@ export default class Wrapper {
+        });
+    }
+
+    static async consumeCode(
+        input:
+            | {
+                  userInputCode: string;
+                  userContext?: any;
+                  options?: RecipeFunctionOptions;
+              }
+            | {
+                  userContext?: any;
+                  options?: RecipeFunctionOptions;
+              }
+    ): Promise<
+        | {
+              status: "OK";
+              createdUser: boolean;
+              user: PasswordlessUser;
+              fetchResponse: Response;
+          }
+        | {
+              status: "INCORRECT_USER_INPUT_CODE_ERROR" | "EXPIRED_USER_INPUT_CODE_ERROR";
+              failedCodeInputAttemptCount: number;
+              maximumCodeInputAttempts: number;
+              fetchResponse: Response;
+          }
+        | { status: "RESTART_FLOW_ERROR"; fetchResponse: Response }
+    > {
+        const recipe: Passwordless = Passwordless.getInstanceOrThrow();
+        const normalisedUserContext = getNormalisedUserContext(input.userContext);
+
+        let additionalParams:
+            | {
+                  userInputCode: string;
+                  deviceId: string;
+                  preAuthSessionId: string;
+              }
+            | {
+                  linkCode: string;
+                  preAuthSessionId: string;
+              };
+
+        if ("userInputCode" in input) {
+            const attemptInfoFromStorage = await recipe.recipeImpl.getLoginAttemptInfo({
+                userContext: normalisedUserContext,
+            });
+
+            /**
+             * If attemptInfoFromStorage is undefined then local storage was probably cleared by another tab.
+             * In this case we use empty strings when calling the API because we want to
+             * return "RESTART_FLOW_ERROR"
+             *
+             * Note: We dont do this for the linkCode flow because that does not depend on local storage.
+             */
+
+            additionalParams = {
+                userInputCode: input.userInputCode,
+                deviceId: attemptInfoFromStorage === undefined ? "" : attemptInfoFromStorage.deviceId,
+                preAuthSessionId: attemptInfoFromStorage === undefined ? "" : attemptInfoFromStorage.preAuthSessionId,
+            };
+        } else {
+            const linkCode = recipe.recipeImpl.getLinkCodeFromURL({
+                userContext: input.userContext,
+            });
+
+            const preAuthSessionId = recipe.recipeImpl.getPreAuthSessionIdFromURL({
+                userContext: input.userContext,
+            });
+
+            additionalParams = {
+                linkCode,
+                preAuthSessionId,
+            };
+        }
+
+        return recipe.recipeImpl.consumeCode({
+            userContext: normalisedUserContext,
+            options: input.options,
+            ...additionalParams,
+        });
```

Comment:
Whilst this is fine, it's being repeated in web-js SDK as well. Is there anyway we can not have this duplicate code across the two SDKs?

I know that we can't just call the the web-js' exposed function here cause maybe the user has overrides the recipe functions of this SDK. But a solution could be that in webjs, we have a file that has all these functions but they all take a recipe interface as an input too. And then those functions use that recipe interface for the logic instead of `Passwordless.getInstanceOrThrow(). recipeImpl `.

Then this SDK can call those functions with `Passwordless.getInstanceOrThrow(). recipeImpl` as an input and the web-js' index file functions can call those functions with its `Passwordless.getInstanceOrThrow(). recipeImpl` as input

Reply:
I dont like the idea of passing `recipeImpl`, but I also cant think of any other way of avoiding duplication without adding some complexity. In this case I think it is fine to leave it as is

Reply:
ok